### PR TITLE
dependency: remove pycurl package

### LINKED ISSUE #30763: [$100 bounty] Remove pycurl usage
It's used in [tools/lib/url_file.py](tools/lib/url_file.py) and [qcomgpsd](system/qcomgpsd/qcomgpsd.py).
* replace current usage with requests or urllib3
* remove `pycurl` from the pyproject.toml

### PR DESCRIPTION
**Pull request by royjr**
may fix #30763
```
comma@comma-33e0a0c6:/data/openpilot$ python tools/lib/tests/test_caching.py
....
----------------------------------------------------------------------
Ran 4 tests in 3.145s

OK
```

### CODE REVIEW

File: system/qcomgpsd/qcomgpsd.py [hunk 1]
```diff
@@ -101,29 +102,23 @@ def gps_enabled() -> bool:
     raise Exception("failed to execute QGPS mmcli command") from exc
 
 def download_assistance():
+  http = urllib3.PoolManager()
```
Comment (adeebshihadeh) [inline system/qcomgpsd/qcomgpsd.py:105]: any reason you switched to urllib? the requests version of this looked a bit nicer

Comment (royjr) [reply system/qcomgpsd/qcomgpsd.py:105]: just PR consistency, but I agree, reverted

File: system/qcomgpsd/qcomgpsd.py [hunk 2]
```diff
@@ -102,27 +103,21 @@ def gps_enabled() -> bool:
...
-      c = pycurl.Curl()
-      c.setopt(pycurl.URL, ASSISTANCE_URL)
-      c.setopt(pycurl.CONNECTTIMEOUT, 5)
-
-      c.setopt(pycurl.WRITEDATA, fp)
-      c.perform()
-      c.close()
-      os.rename(ASSIST_DATA_FILE_DOWNLOAD, ASSIST_DATA_FILE)
-  except pycurl.error:
+      for chunk in response.iter_content(chunk_size=8192):
```
Comment (adeebshihadeh) [inline system/qcomgpsd/qcomgpsd.py:115]: if we're streaming, no need to do the head request first. just check in the loop

File: system/qcomgpsd/qcomgpsd.py [hunk 3]
```diff
@@ -102,27 +103,20 @@ def gps_enabled() -> bool:
...
+        downloaded_size += len(chunk)
+        if downloaded_size > 1e5:
+          cloudlog.error("Qcom assistance data larger than expected")
+          return
+        fp.write(chunk)
+
+    os.rename(ASSIST_DATA_FILE_DOWNLOAD, ASSIST_DATA_FILE)
+
+  except RequestException:
```
Comment (adeebshihadeh) [inline system/qcomgpsd/qcomgpsd.py:119]: use `requests.exceptions.RequestException`

File: system/qcomgpsd/qcomgpsd.py [hunk 4]
```diff
@@ -102,27 +103,20 @@ def gps_enabled() -> bool:
...
-      c.setopt(pycurl.CONNECTTIMEOUT, 5)
-
-      c.setopt(pycurl.WRITEDATA, fp)
-      c.perform()
-      c.close()
-      os.rename(ASSIST_DATA_FILE_DOWNLOAD, ASSIST_DATA_FILE)
-  except pycurl.error:
+      for chunk in response.iter_content(chunk_size=8192):
+        downloaded_size += len(chunk)
+        if downloaded_size > 1e5:
```
Comment (adeebshihadeh) [inline system/qcomgpsd/qcomgpsd.py:112]: can replace this with `fp.tell()`

File: tools/lib/url_file.py [hunk 5]
```diff
@@ -35,13 +36,14 @@ def __init__(self, url, debug=False, cache=None):
...
       self._force_download = not cache
 
-    try:
-      self._curl = self._tlocal.curl
-    except AttributeError:
-      self._curl = self._tlocal.curl = pycurl.Curl()
     if not self._force_download:
       os.makedirs(Paths.download_cache_root(), exist_ok=True)
 
+  def _get_http_client(self):
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:42]: wrong indentation again. can you see why the linter isn't catching this?

Comment (royjr) [reply tools/lib/url_file.py:42]: Changing ruff to..
```
[tool.ruff]
preview = true
select = ["E", "F", "W", "PIE", "C4", "ISC", "RUF008", "RUF100", "A", "B", "TID251", "E117"]
```
should be what were [looking for](https://docs.astral.sh/ruff/rules/over-indented/) but we'd need to set the [indent width](https://docs.astral.sh/ruff/settings/#indent-width) to 2 which wasn't working for me
```
E111 Indentation is not a multiple of 4
```

Comment (adeebshihadeh) [reply tools/lib/url_file.py:42]: can we turn off E111?

File: tools/lib/url_file.py [hunk 6]
```diff
@@ -117,58 +116,42 @@ def read(self, ll=None):
...
-      c.setopt(pycurl.HEADERFUNCTION, header)
-
-      def test(debug_type, debug_msg):
-       print("  debug(%d): %s" % (debug_type, debug_msg.strip()))
-
-      c.setopt(pycurl.VERBOSE, 1)
-      c.setopt(pycurl.DEBUGFUNCTION, test)
       t1 = time.time()
 
-    c.perform()
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:156]: the actual request should be here for the timing to work properly

File: tools/lib/url_file.py [hunk 7]
```diff
@@ -53,17 +55,14 @@ def __exit__(self, exc_type, exc_value, traceback):
...
-    c.reset()
-    return length
+    retries = Retry(connect=3, read=2, redirect=5)
+    timeout = Timeout(connect=5.0, read=50.0)
+    try:
+      response = self._get_http_client().request('HEAD', self._url, retries=retries, timeout=timeout, preload_content=False)
+      length = response.headers.get('content-length', 0)
+      return int(length)
+    except urllib3.exceptions.HTTPError as e:
+      raise e
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:65]: what's the exception handling for here? doesn't seem like it does anything

File: tools/lib/url_file.py [hunk 8]
```diff
@@ -53,17 +55,14 @@ def __exit__(self, exc_type, exc_value, traceback):
...
-    c.setopt(pycurl.TIMEOUT_MS, 500000)
-    c.setopt(pycurl.FOLLOWLOCATION, True)
-    c.setopt(pycurl.URL, self._url)
-    c.setopt(c.NOBODY, 1)
-    c.perform()
-    length = int(c.getinfo(c.CONTENT_LENGTH_DOWNLOAD))
-    c.reset()
-    return length
+    retries = Retry(connect=3, read=2, redirect=5)
+    timeout = Timeout(connect=5.0, read=50.0)
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:59]: looks like this should be 500s and let's make connect 50

File: tools/lib/url_file.py [hunk 9]
```diff
@@ -35,13 +36,14 @@ def __init__(self, url, debug=False, cache=None):
...
-      self._curl = self._tlocal.curl
-    except AttributeError:
-      self._curl = self._tlocal.curl = pycurl.Curl()
     if not self._force_download:
       os.makedirs(Paths.download_cache_root(), exist_ok=True)
 
+  def _get_http_client(self):
+    if not hasattr(self._tlocal, 'http_client'):
+      self._tlocal.http_client = PoolManager()
+    return self._tlocal.http_client
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:45]: why was this moved to a function instead of doing in the constructor?

File: tools/lib/url_file.py [hunk 10]
```diff
@@ -117,58 +116,42 @@ def read(self, ll=None):
...
-      headers.append(f"Range: bytes={self._pos}-{end}")
+      headers['Range'] = f"bytes={self._pos}-{end}"
       download_range = True
 
-    dats = BytesIO()
-    c = self._curl
-    c.setopt(pycurl.URL, self._url)
-    c.setopt(pycurl.WRITEDATA, dats)
-    c.setopt(pycurl.NOSIGNAL, 1)
-    c.setopt(pycurl.TIMEOUT_MS, 500000)
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:136]: looks like this doesn't match?

File: tools/lib/url_file.py [hunk 11]
```diff
@@ -117,58 +116,42 @@ def read(self, ll=None):
...
         return b""
-      headers.append(f"Range: bytes={self._pos}-{end}")
+      headers['Range'] = f"bytes={self._pos}-{end}"
       download_range = True
 
-    dats = BytesIO()
-    c = self._curl
-    c.setopt(pycurl.URL, self._url)
-    c.setopt(pycurl.WRITEDATA, dats)
-    c.setopt(pycurl.NOSIGNAL, 1)
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:135]: and is this handled?

File: tools/lib/url_file.py [hunk 12]
```diff
@@ -1,12 +1,13 @@
 import os
 import time
 import threading
-import pycurl
 from hashlib import sha256
-from io import BytesIO
 from tenacity import retry, wait_random_exponential, stop_after_attempt
 from openpilot.common.file_helpers import atomic_write_in_dir
 from openpilot.system.hardware.hw import Paths
+import urllib3
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:8]: move these to the top. import order is stdlib, new line, openpilot imports

File: tools/lib/url_file.py (hunk 5 above)
Comment (royjr) [reply tools/lib/url_file.py:42]: preview
```
preview = true
Found 17014 errors.
```
```
preview = true
ignore = [.. "E111"]
Found 2630 errors.
```

no preview
```
(overindented - 4 spaces)
ignore = [.. "E111"]
Passed
```
```
(overindented - over 4 spaces)
ignore = [.. "E111"]
Found 1 errors. E117 Over-indented
```

adding this did not make a difference
```
indent-width = 2
```

Comment (adeebshihadeh) [reply tools/lib/url_file.py:42]: ah seems like it's not ready in ruff yet? that's fine if that's the case

File: tools/lib/url_file.py (hunk 11 above)
Comment (royjr) [reply tools/lib/url_file.py:135]: > urllib3 doesn't directly expose an equivalent of NOSIGNAL. It's generally safer and more stable, especially in a multithreaded environment.

File: tools/lib/url_file.py [hunk 13]
```diff
@@ -35,13 +36,13 @@ def __init__(self, url, debug=False, cache=None):
...
       self._force_download = not cache
 
-    try:
-      self._curl = self._tlocal.curl
-    except AttributeError:
-      self._curl = self._tlocal.curl = pycurl.Curl()
     if not self._force_download:
       os.makedirs(Paths.download_cache_root(), exist_ok=True)
 
+    if not hasattr(URLFile._tlocal, 'http_client'):
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:42]: the try/except is more thread safe

Comment (royjr) [reply tools/lib/url_file.py:42]: like this?
```
try:
  self._http_client = URLFile._tlocal.http_client
except AttributeError:
  URLFile._tlocal.http_client = PoolManager()
  self._http_client = URLFile._tlocal.http_client
```

**Comment by royjr** [comment]
fixed debug
```
ValueError: Format specifier missing precision
```
by setting precision value
```
get https://commadataci.blob.core.windows.net/openpilotci/0375fdf7b1ce594d/2019-06-13--08-32-25/3/qlog.bz2 {'Connection': 'keep-alive', 'Range': 'bytes=239824-239923'} 0.284 slow
get https://commadataci.blob.core.windows.net/openpilotci/0375fdf7b1ce594d/2019-06-13--08-32-25/3/qlog.bz2 {'Connection': 'keep-alive'} 0.548 slow
```

File: tools/lib/url_file.py (hunk 13 above)
Comment (adeebshihadeh) [reply tools/lib/url_file.py:42]: why not just copy the way the curl one was done?

File: tools/lib/url_file.py [hunk 14]
```diff
@@ -53,17 +55,11 @@ def __exit__(self, exc_type, exc_value, traceback):
...
-    c.setopt(pycurl.NOSIGNAL, 1)
-    c.setopt(pycurl.TIMEOUT_MS, 500000)
-    c.setopt(pycurl.FOLLOWLOCATION, True)
-    c.setopt(pycurl.URL, self._url)
-    c.setopt(c.NOBODY, 1)
-    c.perform()
-    length = int(c.getinfo(c.CONTENT_LENGTH_DOWNLOAD))
-    c.reset()
-    return length
+    retries = Retry(connect=3, read=2, redirect=5)
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:58]: what are the retries for? was curl doing this?

File: tools/lib/url_file.py [hunk 15]
```diff
@@ -117,58 +113,38 @@ def read(self, ll=None):
...
-
-      def test(debug_type, debug_msg):
-       print("  debug(%d): %s" % (debug_type, debug_msg.strip()))
-
-      c.setopt(pycurl.VERBOSE, 1)
-      c.setopt(pycurl.DEBUGFUNCTION, test)
       t1 = time.time()
 
-    c.perform()
+    retries = Retry(connect=3, read=2, redirect=5)
```
Comment (adeebshihadeh) [inline tools/lib/url_file.py:130]: same here

**Comment by adeebshihadeh** [comment]
Thanks, this was one of the worse dependencies for native macOS support. 

I just bumped up this one to $500 if you wanna try it - https://github.com/commaai/openpilot/issues/30706.
